contracts/RangedMarkets/RangedMarketsAMM.sol

```diff
@@ -0,0 +1,621 @@
+        uint leftQuote,
+        uint rightQuote,
+        uint additionalSlippage
+    ) internal {
+        thalesAmm.sellToAMM(
+            address(rangedMarket.leftMarket()),
+            IThalesAMM.Position.Up,
+            amount / 2,
+            leftQuote,
+            additionalSlippage
+        );
+
+        thalesAmm.sellToAMM(
+            address(rangedMarket.rightMarket()),
+            IThalesAMM.Position.Down,
+            amount / 2,
+            rightQuote,
+            additionalSlippage
+        );
+    }
+
+    function _updateSpentOnMarketAndSafeBoxOnSell(
+        uint amount,
+        RangedMarket rangedMarket,
+        uint sUSDPaid
+    ) internal {
+        uint safeBoxShare = ((sUSDPaid * ONE) / (ONE - safeBoxImpact)) - sUSDPaid;
+
+        if (safeBoxImpact > 0) {
+            sUSD.transfer(safeBox, safeBoxShare);
+        } else {
+            safeBoxShare = 0;
+        }
+
+        // consider that amount might be 18 decimals
+        if (amount > (spentOnMarket[address(rangedMarket)] + sUSDPaid + safeBoxShare)) {
+            spentOnMarket[address(rangedMarket)] = 0;
+        } else {
+            spentOnMarket[address(rangedMarket)] = spentOnMarket[address(rangedMarket)] + sUSDPaid + safeBoxShare - amount;
+        }
+    }
+
+    function withdrawCollateral(RangedMarket rangedMarket)
+        external
+        onlyOwner
+        knownRangedMarket(address(rangedMarket))
+        nonReentrant
+        notPaused
+    {
+        rangedMarket.withdrawCollateral(safeBox);
+    }
+
+    function setRangedMarketMastercopy(address _rangedMarketMastercopy) external onlyOwner {
+        rangedMarketMastercopy = _rangedMarketMastercopy;
+    }
+
+    function setRangedPositionMastercopy(address _rangedPositionMastercopy) external onlyOwner {
+        rangedPositionMastercopy = _rangedPositionMastercopy;
+    }
+
+    function setMinMaxSupportedPrice(
+        uint _minSupportedPrice,
+        uint _maxSupportedPrice,
+        uint _minDiffBetweenStrikes
+    ) public onlyOwner {
+        minSupportedPrice = _minSupportedPrice;
+        maxSupportedPrice = _maxSupportedPrice;
+        minimalDifBetweenStrikes = _minDiffBetweenStrikes;
+        emit SetMinSupportedPrice(minSupportedPrice);
+        emit SetMaxSupportedPrice(maxSupportedPrice);
+        emit SetMinimalDifBetweenStrikes(minimalDifBetweenStrikes);
+    }
+
+    function setSafeBoxData(address _safeBox, uint _safeBoxImpact) public onlyOwner {
+        safeBoxImpact = _safeBoxImpact;
+        safeBox = _safeBox;
+        emit SetSafeBoxImpact(_safeBoxImpact);
+        emit SetSafeBox(_safeBox);
+    }
+
+    function setCapPerMarketAndRangedAMMFee(uint _capPerMarket, uint _rangedAMMFee) public onlyOwner {
+        capPerMarket = _capPerMarket;
+        rangedAmmFee = _rangedAMMFee;
+        emit SetCapPerMarket(capPerMarket);
+        emit SetRangedAmmFee(rangedAmmFee);
+    }
+
+    function setThalesAMM(address _thalesAMM) public onlyOwner {
```

Comment:
maybe to check zero address on this AMM address, it should not be zero i guess

Reply:
I kept running out of contract size, and since these are admin methods, I thoughts its fine to not have those kinds of checks.